File: MergeIntervals.py
# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e


class Interval:
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e
    def __repr__(self):
        return str(self.start) + " "+ str(self.end)

class Solution:
    # @param intervals, a list of Interval
    # @return a list of Interval
    # Wrong ! should return list or intervals
    def merge(self, intervals):
        if intervals == []:
            return []
            # Sort according to start
        res = []
        intervals.sort(key=lambda t: t.start)
            
        start, end = intervals[0].start, intervals[0].end
        for i in range(len(intervals)):
            if intervals[i].end < end:
                continue
            # interevals[i][1] >= end (current)
            # concatenate
            if intervals[i].start <= end:
                end = intervals[i].end
            else: # intervals[i][0] > end (a gap)
                res.append([start, end])
                start, end = intervals[i].start, intervals[i].end
                    
        res.append([start,end])
        return res
    
    # Right
    def merge2(self, intervals):
        if intervals == []:
            return []
        intervals.sort(key=lambda a: a.start)
        res = []
        for i in intervals:
            # the first or not overlapped
            if len(res) == 0 or res[-1].end < i.start:
                res.append(i)
            else:
                x = res[-1]
                x.start = min(x.start, i.start)
                x.end = max(x.end, i.end)
        return res

if __name__ == '__main__':
    # i1 = Interval(1,3)
    # i2 = Interval(2,6)
    # i3 = Interval(8,10)
    # i4 = Interval(15,18)

    # l = [i1,i2,i3,i4]
    so = Solution()
    # print so.merge(l)
    ll = [Interval(1, 4), Interval(0, 0)]
    res = so.merge2(ll)
    for i in res:
        print i.start, i.end

File: CPP/Second/merge_intervals.cc
// https://oj.leetcode.com/problems/merge-intervals/

// Given a collection of intervals, merge all overlapping intervals.
// For example,
// Given [1,3],[2,6],[8,10],[15,18],
// return [1,6],[8,10],[15,18].

/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */

class Solution {
public:
    vector<Interval> merge(vector<Interval> &intervals) {
        /*
        iterate each Interval, there are two cases.
        #1, not overlapped, there will be a gap, then add it
        #2, overlapped, update the last one
        */
        sort(intervals.begin(), intervals.end(), 
            [](const Interval &a, const Interval &b) {return a.start < b.start;});
        vector<Interval> res;
        for (auto i: intervals) {
            if (res.empty() || res.back().end < i.start) {
                res.push_back(i);
            }
            else {
                Interval &x = res.back();
                x.start = min(i.start, x.start);
                x.end = max(i.end, x.end);
            }
        }
        return res;
    }
};

int main(int argc, char *argv[])
{

    return 0;
}
